Add template store command

martindekov opened the pull request:
Moving template command as own command which has store and
pull commands as subcommands. Also adding store list and
store pull commands to list and pull from store

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Also opened Issue in store which slaps the keys: [43](https://github.com/openfaas/store/pull/44)

This PR separates `template pull` command into separate commands and adding additional `template store list` and `template store pull` for interactions with the official templates store.

The commands are structured the following way:

* `template`

    * `pull`

    * `store`

        * `list`
        
        * `pull`

And are used as before, there are the ways you can use every command now:

`faas-cli template pull <URL OF REPOSITORY>`' <- (default)
`faas-cli template pull <URL OF REPOSITORY> --overwrite`
`faas-cli template pull <URL OF REPOSITORY> --debug`
`faas-cli template store list`
`faas-cli template store ls`
`faas-cli template store list --verbose`
`faas-cli template store list --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json`
The fields inside `<>` aka `<NAME>` is the content below the fields seen through the command above aka.
```
faas-cli template store list

NAME      SOURCE      DESCRIPTION
...       ...         ...
```

`faas-cli template store pull <NAME>||<SOURCE/NAME> ...` <- (default)
`faas-cli template store pull <NAME>||<SOURCE/NAME> ... --overwrite`
`faas-cli template store pull <NAME>||<SOURCE/NAME> ... --debug`
`faas-cli template store pull <NAME>||<SOURCE/NAME> ... --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json`

Also you can set the environmental variable `OPENFAAS_TEMPLATE_STORE_URL` which will prioritize over the default one the following way `argument` > `environmental_variable` > `default`

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Closes https://github.com/openfaas/faas-cli/issues/559

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Manual tests:

`faas-cli template store ls`
```
> ./faas-cli template store ls

NAME                    SOURCE             DESCRIPTION
csharp                  openfaas           Official C# template
...
node10-express-armhf    openfaas-incubator NodeJS 10 Express armhf template
...
```

`faas-cli template store ls -v`
```
> ./faas-cli template store ls -v

NAME                    LANGUAGE   PLATFORM SOURCE             DESCRIPTION
csharp                  C#         x86_64   openfaas           Official C# template
...
node10-express-armhf    NodeJS     armhf    openfaas-incubator NodeJS 10 Express armhf template
...
```

`faas-cli template store pull csharp golang-http python3-debian`
```
> ./faas-cli template store pull csharp golang-http python3-debian
Fetch templates from repository: https://github.com/openfaas/templates
2018/11/26 01:47:35 Attempting to expand templates from https://github.com/openfaas/templates
2018/11/26 01:47:37 Fetched 14 template(s) : [csharp dockerfile go go-armhf java8 node node-arm64 node-armhf php7 python python-armhf python3 python3-armhf ruby] from https://github.com/openfaas/templates
Fetch templates from repository: https://github.com/openfaas-incubator/golang-http-template
2018/11/26 01:47:37 Attempting to expand templates from https://github.com/openfaas-incubator/golang-http-template
2018/11/26 01:47:38 Fetched 4 template(s) : [golang-http golang-http-armhf golang-middleware golang-middleware-armhf] from https://github.com/openfaas-incubator/golang-http-template
Fetch templates from repository: https://github.com/openfaas-incubator/python3-debian
2018/11/26 01:47:38 Attempting to expand templates from https://github.com/openfaas-incubator/python3-debian
2018/11/26 01:47:40 Fetched 1 template(s) : [python3-debian] from https://github.com/openfaas-incubator/python3-debian
```
`faas-cli template store pull csharp what golang-http template python3-debian --overwrite`
```
> ./faas-cli template store pull csharp what golang-http template python3-debian --overwrite
Fetch templates from repository: https://github.com/openfaas/templates
2018/11/26 01:48:19 Attempting to expand templates from https://github.com/openfaas/templates
2018/11/26 01:48:20 Fetched 14 template(s) : [csharp dockerfile go go-armhf java8 node node-arm64 node-armhf php7 python python-armhf python3 python3-armhf ruby] from https://github.com/openfaas/templates
Fetch templates from repository: https://github.com/openfaas-incubator/golang-http-template
2018/11/26 01:48:20 Attempting to expand templates from https://github.com/openfaas-incubator/golang-http-template
2018/11/26 01:48:22 Fetched 4 template(s) : [golang-http golang-http-armhf golang-middleware golang-middleware-armhf] from https://github.com/openfaas-incubator/golang-http-template
Fetch templates from repository: https://github.com/openfaas-incubator/python3-debian
2018/11/26 01:48:22 Attempting to expand templates from https://github.com/openfaas-incubator/python3-debian
2018/11/26 01:48:23 Fetched 1 template(s) : [python3-debian] from https://github.com/openfaas-incubator/python3-debian

There is no template with name: `what` in the store.

There is no template with name: `template` in the store.

check available options by running:

faas-cli template store pull <NAME>|<SOURCE/NAME>

```
`faas-cli template store pull `
```
> ./faas-cli template store pull 

Need to specify one of the store templates check available ones by running the command:

faas-cli template store list

```
`faas-cli template store pull go --url=asd`
```
> ./faas-cli template store pull go --url=asd
Error while fetching templates from store: error while requesting template list: Get asd: unsupported protocol scheme ""
```
`faas-cli template store list go --url=asd`
```
> ./faas-cli template store list go --url=asd
Error while getting templates info: error while requesting template list: Get asd: unsupported protocol scheme ""
```
`faas-cli template store pull golang-http python3-debian --debug`
```
> ./faas-cli template store pull golang-http python3-debian --debug
Fetch templates from repository: https://github.com/openfaas-incubator/golang-http-template
2018/11/26 01:58:01 Attempting to expand templates from https://github.com/openfaas-incubator/golang-http-template
Temp files in /tmp/openFaasTemplates879867391
2018/11/26 01:58:04 Cannot overwrite the following 4 template(s): [golang-http golang-http-armhf golang-middleware golang-middleware-armhf]
2018/11/26 01:58:04 Fetched 0 template(s) : [] from https://github.com/openfaas-incubator/golang-http-template
Fetch templates from repository: https://github.com/openfaas-incubator/python3-debian
2018/11/26 01:58:04 Attempting to expand templates from https://github.com/openfaas-incubator/python3-debian
Temp files in /tmp/openFaasTemplates578810194
2018/11/26 01:58:06 Cannot overwrite the following 1 template(s): [python3-debian]
2018/11/26 01:58:06 Fetched 0 template(s) : [] from https://github.com/openfaas-incubator/python3-debian
```
`faas-cli template store list --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json`
```
> ./faas-cli template store list --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json

NAME                    SOURCE             DESCRIPTION
csharp                  openfaas           Official C# template
...
node10-express-armhf    openfaas-incubator NodeJS 10 Express armhf template
...
```

`faas-cli template store list --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json -v`

```
> ./faas-cli template store list --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json -v

NAME                    LANGUAGE   PLATFORM SOURCE             DESCRIPTION
csharp                  C#         x86_64   openfaas           Official C# template
...
node10-express-armhf    NodeJS     armhf    openfaas-incubator NodeJS 10 Express armhf template
...
```

`faas-cli template store pull csharp ruby-http python27-flask --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json --overwrite`

```
> ./faas-cli template store pull csharp ruby-http python27-flask --url https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json --overwrite
Fetch templates from repository: https://github.com/openfaas/templates
2018/11/26 02:04:28 Attempting to expand templates from https://github.com/openfaas/templates
2018/11/26 02:04:30 Fetched 14 template(s) : [csharp dockerfile go go-armhf java8 node node-arm64 node-armhf php7 python python-armhf python3 python3-armhf ruby] from https://github.com/openfaas/templates
Fetch templates from repository: https://github.com/openfaas-incubator/ruby-http
2018/11/26 02:04:30 Attempting to expand templates from https://github.com/openfaas-incubator/ruby-http
2018/11/26 02:04:31 Fetched 1 template(s) : [ruby-http] from https://github.com/openfaas-incubator/ruby-http
Fetch templates from repository: https://github.com/openfaas-incubator/python-flask-template
2018/11/26 02:04:31 Attempting to expand templates from https://github.com/openfaas-incubator/python-flask-template
2018/11/26 02:04:33 Fetched 3 template(s) : [python27-flask python3-flask python3-flask-armhf] from https://github.com/openfaas-incubator/python-flask-template
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

ivanayov commented:
I tested out and looks good. 

Can you add also `faas-cli template store ls`? It's the common usage of `list` and is the intuitive way for most people.

I suggest `faas-cli template store list --verbose` to give the url as well.

I tested this command and the output seems not to be quite applicable. May be it's a copy-paste issue.
```
$ ./faas-cli template store pull --help
List all templates supported by openfaas and openfaas-incubator organizations

Usage:
  faas-cli template store pull [flags]

Examples:
faas-cli template store pull <>
```

martindekov commented:
Thank you for the feedback @ivanayov for the `help` I am aware that this is unfinished I wanted to make the commands work and the details I will ammend tommorow

martindekov commented on commands/template_pull_test.go:
```diff
@@ -62,16 +61,17 @@ func Test_templatePull(t *testing.T) {
+			var buf bytes.Buffer
+			faasCmd.SetArgs([]string{"template", "pull", "user@host.xz:openfaas/faas-cli.git"})
+			faasCmd.SetOutput(&buf)
+			err := faasCmd.Execute()
```
Test fails here and exits with failing, not sure why the buffer is not populated by the output

martindekov commented on commands/template_pull_test.go:
```diff
@@ -3,15 +3,13 @@
 	"testing"
 )
 
+/*
```
@rgee0 If you can take a look I will appreciate that big time. Not sure why tests fail, but when I run the command by hand they pass ..

martindekov replied:
This is the output
```
--- FAIL: Test_templatePull (0.03s)
    --- FAIL: Test_templatePull/ValidRepo (0.00s)
    	/home/martindekov/go/src/github.com/openfaas/faas-cli/commands/template_pull_test.go:27: The directory templates was not created
    	/home/martindekov/go/src/github.com/openfaas/faas-cli/commands/fetch_templates_test.go:70: Directory template was not created: stat template/: no such file or directory
    --- FAIL: Test_templatePull/WithOverwriting (0.00s)
    	/home/martindekov/go/src/github.com/openfaas/faas-cli/commands/template_pull_test.go:46: 
    	/home/martindekov/go/src/github.com/openfaas/faas-cli/commands/fetch_templates_test.go:70: Directory template was not created: stat template/: no such file or directory
```

martindekov commented on commands/template_pull_test.go:
```diff
@@ -62,16 +61,20 @@ func Test_templatePull(t *testing.T) {
-		}
-	})
+	/*
+		t.Run("InvalidUrlError", func(t *testing.T) {
```
Sorry my mistake the above ones were failing ... this was the problematic test @rgee0 the output is
```

Fetch templates from repository: /tmp/openFaasTestTemplates319700903
2018/11/22 06:57:09 Attempting to expand templates from /tmp/openFaasTestTemplates319700903
2018/11/22 06:57:09 Fetched 2 template(s) : [dockerfile ruby] from /tmp/openFaasTestTemplates319700903
Fetch templates from repository: /tmp/openFaasTestTemplates319700903
2018/11/22 06:57:09 Attempting to expand templates from /tmp/openFaasTestTemplates319700903
2018/11/22 06:57:09 Fetched 2 template(s) : [dockerfile ruby] from /tmp/openFaasTestTemplates319700903
Fetch templates from repository: /tmp/openFaasTestTemplates319700903
Fetch templates from repository: /tmp/openFaasTestTemplates319700903
The repository URL must be a valid git repo uri
FAIL	github.com/openfaas/faas-cli/commands	0.106s
Error: Tests failed.
```
I see the output contains what I need I don't know though why it is in that format

martindekov replied:
`os.Exit(1)` was the problem in the situation

rgee0 commented:
```
$ git checkout templateStore
Switched to branch 'templateStore'
$ ./build_redist.sh

Sending build context to Docker daemon  51.01MB
Step 1/17 : FROM golang:1.10.4 as builder
 ---> a4afc24299ee
Step 2/17 : WORKDIR /usr/bin/
 ---> Using cache
 ---> 90f41a4d9516
Step 3/17 : RUN curl -sLSf https://raw.githubusercontent.com/alexellis/license-check/maste
r/get.sh | sh
 ---> Using cache
 ---> 41f323f6f8ee
Step 4/17 : WORKDIR /go/src/github.com/openfaas/faas-cli
 ---> Using cache
 ---> 9ca060d09fb2
Step 5/17 : COPY . .
 ---> 8f2fb594a06d
Step 6/17 : RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*")
)" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
 ---> Running in 08c5c5bdc05c
Removing intermediate container 08c5c5bdc05c
 ---> d904d83c48cb
Step 7/17 : RUN /usr/bin/license-check -path ./ --verbose=false "Alex Ellis" "OpenFaaS Aut
hor(s)"  && go test $(go list ./... | grep -v /vendor/ | grep -v /template/|grep -v /build
/) -cover  && VERSION=$(git describe --all --exact-match `git rev-parse HEAD` | grep tags
| sed 's/tags\///')  && GIT_COMMIT=$(git rev-list -1 HEAD)  && CGO_ENABLED=0 GOOS=linux go
 build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_CO
MMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -inst
allsuffix cgo -o faas-cli  && CGO_ENABLED=0 GOOS=darwin go build --ldflags "-s -w
-X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/open
faas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli-darwin
  && CGO_ENABLED=0 GOOS=windows go build --ldflags "-s -w         -X github.com/openfaas/f
aas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Ve
rsion=${VERSION}"         -a -installsuffix cgo -o faas-cli.exe  && CGO_ENABLED=0 GOOS=lin
ux GOARCH=arm GOARM=6 go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/ve
rsion.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VE
RSION}"         -a -installsuffix cgo -o faas-cli-armhf  && CGO_ENABLED=0 GOOS=linux GOARC
H=arm64 go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommi
t=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"
  -a -installsuffix cgo -o faas-cli-arm64
 ---> Running in 910872f7ba89
ok      github.com/openfaas/faas-cli    0.010s  coverage: 89.4% of statements
ok      github.com/openfaas/faas-cli/builder    0.020s  coverage: 29.6% of statements
ok      github.com/openfaas/faas-cli/commands   4.002s  coverage: 47.7% of statements
ok      github.com/openfaas/faas-cli/config     0.007s  coverage: 81.0% of statements
ok      github.com/openfaas/faas-cli/proxy      0.046s  coverage: 73.3% of statements
?       github.com/openfaas/faas-cli/sample/github_hmac [no test files]
ok      github.com/openfaas/faas-cli/schema     0.015s  coverage: 100.0% of statements
ok      github.com/openfaas/faas-cli/stack      0.028s  coverage: 50.0% of statements
?       github.com/openfaas/faas-cli/test       [no test files]
ok      github.com/openfaas/faas-cli/version    0.004s  coverage: 100.0% of statements
?       github.com/openfaas/faas-cli/versioncontrol     [no test files]
Removing intermediate container 910872f7ba89
 ---> d22ea6a90dbd
Step 8/17 : FROM alpine:3.7
 ---> 3fd9065eaf02
Step 9/17 : RUN apk --no-cache add ca-certificates git
 ---> Using cache
 ---> 8b879f296f81
Step 10/17 : WORKDIR /root/
 ---> Using cache
 ---> 653d85c9293e
Step 11/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli
    .
 ---> 961ab0b6c4f7
Step 12/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-darwin
    .
 ---> 141e646041ac
Step 13/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-armhf
    .
 ---> fc0063227efc
Step 14/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli.exe
    .
 ---> e87a21605d02
Step 15/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-arm64
    .
 ---> de6072076949
Step 16/17 : ENV PATH=$PATH:/root/
 ---> Running in 11794a3934ac
Removing intermediate container 11794a3934ac
 ---> 4db549f2c8f5
Step 17/17 : CMD ["faas-cli"]
 ---> Running in 0a2b74470e8b
Removing intermediate container 0a2b74470e8b
 ---> dd8268401d29
Successfully built dd8268401d29
Successfully tagged openfaas/faas-cli:latest-dev
edddae02e2cbe7f79c4d90cd735098aa94678b4632736a81dc1522ee92933499
faas-cli
```

rgee0 commented:
```
$ git checkout master
Switched to branch 'master'
$ ./build_redist.sh
Sending build context to Docker daemon  58.07MB

Step 1/17 : FROM golang:1.10.4 as builder
 ---> a4afc24299ee
Step 2/17 : WORKDIR /usr/bin/
 ---> Using cache
 ---> 90f41a4d9516
Step 3/17 : RUN curl -sLSf https://raw.githubusercontent.com/alexellis/license-check/master/get.sh | sh
 ---> Using cache
 ---> 41f323f6f8ee
Step 4/17 : WORKDIR /go/src/github.com/openfaas/faas-cli
 ---> Using cache
 ---> 9ca060d09fb2
Step 5/17 : COPY . .
 ---> 934dacdaaccb
Step 6/17 : RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
 ---> Running in 74b9b8aa1511
Removing intermediate container 74b9b8aa1511
 ---> 9b7ec86fccc5
Step 7/17 : RUN /usr/bin/license-check -path ./ --verbose=false "Alex Ellis" "OpenFaaS Author(s)"  && go test $(go list ./... | grep -v /vendor/ | grep -v /template/|grep -v /build/) -cover  && VERSION=$(git describe --all --exact-match `git rev-parse HEAD` | grep tags | sed 's/tags\///')  && GIT_COMMIT=$(git rev-list -1 HEAD)  && CGO_ENABLED=0 GOOS=linux go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli  && CGO_ENABLED=0 GOOS=darwin go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli-darwin  && CGO_ENABLED=0 GOOS=windows go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli.exe  && CGO_ENABLED=0 GOOS=linux GOARCH=arm GOARM=6 go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli-armhf  && CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build --ldflags "-s -w         -X github.com/openfaas/faas-cli/version.GitCommit=${GIT_COMMIT}         -X github.com/openfaas/faas-cli/version.Version=${VERSION}"         -a -installsuffix cgo -o faas-cli-arm64
 ---> Running in a19ed629a8e2
ok  	github.com/openfaas/faas-cli	0.006s	coverage: 89.4% of statements
ok  	github.com/openfaas/faas-cli/builder	0.013s	coverage: 29.6% of statements
ok  	github.com/openfaas/faas-cli/commands	2.539s	coverage: 50.2% of statements
ok  	github.com/openfaas/faas-cli/config	0.036s	coverage: 81.0% of statements
ok  	github.com/openfaas/faas-cli/proxy	0.094s	coverage: 73.3% of statements
?   	github.com/openfaas/faas-cli/sample/github_hmac	[no test files]
ok  	github.com/openfaas/faas-cli/schema	0.017s	coverage: 100.0% of statements
ok  	github.com/openfaas/faas-cli/stack	0.019s	coverage: 50.0% of statements
?   	github.com/openfaas/faas-cli/test	[no test files]
ok  	github.com/openfaas/faas-cli/version	0.003s	coverage: 100.0% of statements
?   	github.com/openfaas/faas-cli/versioncontrol	[no test files]
Removing intermediate container a19ed629a8e2
 ---> 1bf83b8dd097
Step 8/17 : FROM alpine:3.7
 ---> 3fd9065eaf02
Step 9/17 : RUN apk --no-cache add ca-certificates git
 ---> Using cache
 ---> 8b879f296f81
Step 10/17 : WORKDIR /root/
 ---> Using cache
 ---> 653d85c9293e
Step 11/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli                .
 ---> e600c3f59e25
Step 12/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-darwin         .
 ---> b06bea032ae3
Step 13/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-armhf          .
 ---> 758d6cb15e18
Step 14/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli.exe            .
 ---> 760a7bdaf43b
Step 15/17 : COPY --from=builder /go/src/github.com/openfaas/faas-cli/faas-cli-arm64          .
 ---> 948ac6f4ec50
Step 16/17 : ENV PATH=$PATH:/root/
 ---> Running in 01b3d0c6d299
Removing intermediate container 01b3d0c6d299
 ---> b6ee4a9cc590
Step 17/17 : CMD ["faas-cli"]
 ---> Running in 76002c942ab6
Removing intermediate container 76002c942ab6
 ---> 4bac7155aeb1
Successfully built 4bac7155aeb1
Successfully tagged openfaas/faas-cli:latest-dev
e1fed519227f38e115905e3b0ddb8ead3868f37df65cdc82af81d0e3ce712548
faas-cli

rgee0 commented:
@martindekov not sure what your issue is here.  I can build from both your branch and master with all tests passing

martindekov commented:
Yeah @rgee0 I swapped 
```
	if _, err := os.Stat(repository); err != nil {
		var validURL = regexp.MustCompile(gitRemoteRepoRegex)
		if !validURL.MatchString(repository) {
			fmt.Println("The repository URL must be a valid git repo uri")

			os.Exit(1)
		}
	}
```

instead of `os.Exit(1)` to be `return` you can see it in the code if I return that to `os.Exit(1)` it does not complete the test it only exits, and I know the cli is needed to exit with 1 there were some issues around that

martindekov commented on commands/template_pull_test.go:
```diff
@@ -62,16 +61,20 @@ func Test_templatePull(t *testing.T) {
+
+			faasCmd.SetArgs([]string{"template", "pull", "user@host.xz:openfaas/faas-cli.git"})
+			faasCmd.SetOutput(&buf)
+			err := faasCmd.Execute()
```
command returns error instead of `os.Exit(1)` now and the tests do not fail

martindekov commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,169 @@
+
+const (
+	// The URL where the store resides
+	DefaultTemplateList = "https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json"
```
This is still relevant ^

alexellis replied:
@martindekov please do this now or when you're available next.

martindekov replied:
I did it but still uses `cli-lang` until the store PR is merged

martindekov commented:
Hey there team @openfaas/cli could someone review this PR for me? Thanks in advance!

martindekov commented:
Thank you for the feedback @rgee0 now amended changes with `RunE` not sure why it was `os.Exit(1)`.

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+const (
+	// The URL where the store resides
+	DefaultTemplateList = "https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json"
+	incubatorRepo       = "openfaas-incubator"
```
Why do we have this?

martindekov replied:
it names the source of our templates, because including them all every single time might be little bit spammy

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+	}
+	for _, template := range templates {
+		if source != allTemplates {
+			if source == incubatorRepo {
```
Not sure why we need these three code blocks? I'd expect it to be either a named language, a named repo or `all`

martindekov replied:
This is because the sources are `openfaas` and `openfaas-incubator` by default we set all, but it is a little bit spam, since I expect people to be used in the uncubator so this is a way to filter from where are the templates displayed. Like I want `openfaas-incubator` templates only so i will `--source=openfaas-incubator`

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+		if source != allTemplates {
+			if source == incubatorRepo {
+				if template.Source == officialRepo {
+					continue
```
I don't like `continue` used in this way, I'm finding the flow confusing.

martindekov replied:
Yeah ok I will find another way

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+		} else {
+			if verbose {
+				fmt.Fprintf(lineWriter, "%s\t%s\t%s\t%s\t%s\n",
+					template.CliLanguage,
```
Seems like we have a lot of repetition in this function

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+	return buff.String()
+}
+
+type TemplatesInfo []TemplateInfo
```
Not sure we need this type.

martindekov replied:
Yeah we don't really I will remove it I though it is more verbose

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+
+type TemplatesInfo []TemplateInfo
+
+type TemplateInfo struct {
```
TemplateInfo or `TemplateStoreInfo`?

LucasRoesler replied:
We should definitely add docs for each of these public types. Once the strings are there it will also be easier to discuss the correct naming

alexellis replied:
Ok so a comment for the type. See other examples in the code for how to do this.

martindekov replied:
Ok I will swap with `TemplateStoreInfo` and will extend documentation

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+type TemplatesInfo []TemplateInfo
+
+type TemplateInfo struct {
+	CliLanguage string `json:"cli-lang"`
```
Should be `CLILanguage` but I'm not convinced on the naming. Can we change it to something else?

martindekov replied:
I am a straight forward guy 😄 , proposals are welcome

martindekov replied:
Went with `TemplateName` I think it is better that way

alexellis commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,66 @@
+				sourceName := fmt.Sprintf("%s/%s", storeTemplate.Source, storeTemplate.CliLanguage)
+				if template == storeTemplate.CliLanguage || template == sourceName {
+					if overwrite && pullDebug {
+						faasCmd.SetArgs([]string{"template", "pull", storeTemplate.Repository, "--overwrite", "--debug"})
```
What is `--debug`?

martindekov replied:
The `template pull` has this option I guess if there is a problem with the template, I really don't know it was there I though someone would know how to use it so I added it

martindekov replied:
I removed those now, I render them unnecessary in our situation since we pull from secure source(the store)

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+	} else {
+		fmt.Fprintf(lineWriter, "NAME\tSOURCE\tDESCRIPTION\n")
+	}
+	for _, template := range templates {
```
I think we could reduce the number of blocks/levels here by breaking this into a couple steps (1) filter the list, then (2) print the list.  This results in two loops.  Alternatively or at the same time, we could also move some of these blocks to new methods. This will also improve the readability and some repetition.  This loop could look something like 

```golang
for _, template := range templates {
    if skipSource(template, source) {
        continue
    }
    printTemplate(writer, template, verbose)
}
```

martindekov replied:
Great Idea, thanks for the suggestion

martindekov replied:
Also done ^ moved the parts into separate methods (they still have repetition, I cannot escape that with the source ...)

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,170 @@
+
+	formattedOutput := formatTemplatesOutput(templatesInfo)
+
+	fmt.Println(formattedOutput)
```
I would fprint this to `cmd.OutOrStdout()` this returns a writer, typically stdout, but it makes it easier to test the cmd output.

martindekov replied:
Yeah seems perfect, haven't though of it that way

martindekov replied:
This is now done ^

LucasRoesler commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,66 @@
+						faasCmd.SetArgs([]string{"template", "pull", storeTemplate.Repository, "--overwrite", "--debug"})
+					} else if overwrite && !pullDebug {
+						faasCmd.SetArgs([]string{"template", "pull", storeTemplate.Repository, "--overwrite"})
+					} else if !overwrite && pullDebug {
```
Is this style of the CLI calling itself something we want to do or should the method inside of the "template pull" be extracted and exposed? The only other place we do something like this is in tests.

martindekov replied:
Alex was more for moving the method out and mentioned briefly that I will do it that way and leave for the community to give me feedback on the approach, I also think the method should be extracted and applied to this command, I can totally do that

martindekov replied:
Thanks for the suggestion @LucasRoesler , I amended fix which references that, now I don't use the command calling itself, but instead the pull function

martindekov commented:
Of course I will add documentation as requested in both places.

There is some some work left on this PR(extracting the pull method and add info in README), which if I don't have time to finish today will do it in the weekend.

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+		return nil, fmt.Errorf("error while requesting template list: %s\n", clientErr.Error())
+	}
+
+	if res != nil {
```
Should be checking res.Body for nil

martindekov replied:
done ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+		fmt.Printf("\n`The --source` flag needs to have value %s, %s or %s\n", allTemplates, incubatorRepo, officialRepo)
+		return ""
+	}
+	templaetsOutput := formatTemplatesOutput(templates, verbose, source)
```
Typo on this line

martindekov replied:
fixed ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+
+func formatCommandOutput(templates []TemplateInfo) string {
+	if source != allTemplates && source != incubatorRepo && source != officialRepo {
+		fmt.Printf("\n`The --source` flag needs to have value %s, %s or %s\n", allTemplates, incubatorRepo, officialRepo)
```
I am not sure we need a source flag at all?

martindekov replied:
removed the flag ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+	return templaetsOutput
+}
+
+type TemplateInfo struct {
```
Add godoc please as per comment from Lucas

martindekov replied:
I don't get that part, added documentation on the command in [openfaas/docs:90](https://github.com/openfaas/docs/pull/90) if that is what you ment

martindekov replied:
Maybe you mean https://godoc.org/github.com/openfaas/faas-cli/ ? Is there a repository for this? How to modify it ?

martindekov replied:
Ok added comments above the main cmds like
```
...
// templateStoreCmd allows access to pull and list commands from store
var templateStoreCmd = &cobra.Command{
...
```

also added this text in every command file for example in `template_store.go` I added
```
// Copyright (c) OpenFaaS Author(s) 2018. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package commands
...
```

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+type TemplateInfo struct {
+	TemplateName string `json:"cli-lang"`
```
Not keen on cli-lang as a property

rgee0 replied:
The challenge here is that the `--lang` flag started as a truncation of 'language'.  However, as this moves more and more away from a designating a language to something specific to OpenFaaS and the CLI, making the attribute more generic makes what it refers to less obvious.

Perhaps we could use `template` but we don't refer to template on the CLI, we refer to a language.  However, names like `golang-http-armhf` refer to a OF template, rather than a language.

martindekov replied:
Check out thus PR [openfaas/store:45](https://github.com/openfaas/store/pull/44) ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+func formatBasicOutput(lineWriter *tabwriter.Writer, templates []TemplateInfo, source string) {
+	fmt.Fprintf(lineWriter, "NAME\tSOURCE\tDESCRIPTION\n")
+	for _, template := range templates {
+		if source == incubatorRepo && template.Source == incubatorRepo {
```
Why do we need these three conditions?

martindekov replied:
source again will remove

martindekov replied:
source removed ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+func formatVerboseOutput(lineWriter *tabwriter.Writer, templates []TemplateInfo, source string) {
+	fmt.Fprintf(lineWriter, "NAME\tLANGUAGE\tPLATFORM\tSOURCE\tDESCRIPTION\n")
+	for _, template := range templates {
+		if source == incubatorRepo && template.Source == incubatorRepo {
```
Not sure why we need these three conditions.   What do they do and why?

martindekov replied:
source removed ✔️

alexellis commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+const (
+	defaultTemplateList = "https://raw.githubusercontent.com/martindekov/store/martindekov/unofficial_templates_json/templates.json"
```
Needs to change

martindekov replied:
changed ✔️

alexellis commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,61 @@
+
+var templateStorePullCmd = &cobra.Command{
+	Use:   `pull [TEMPLATE_NAME]`,
+	Short: `Pull templates from store`,
```
I assume we have a flag to override the store json file location right?

martindekov replied:
Added the `--url` flag here also ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,175 @@
+	Use:     `list`,
+	Short:   `List templates from OpenFaaS organizations`,
+	Aliases: []string{"ls"},
+	Long:    `List templates supported by openfaas or openfaas-incubator organizations or both`,
```
How do I specify the store url like with the function store commands?

alexellis replied:
Is this a miss?

martindekov replied:
I added `--url` flag for both `pull` and `list` commands ✔️

alexellis commented on commands/template_store.go:
```diff
@@ -0,0 +1,21 @@
+	Long: `This command provides the list of the templates from the store
+	 that we support and pull them by name`,
+	Example: `  faas-cli template store list --verbose
+  faas-cli template store ls -v
```
Please remove the source flag. This was not part of the design. If anything the part we need has been missed which is to provide a custom URL for a template store.

martindekov replied:
source removed ✔️

alexellis commented on README.md:
```diff
@@ -145,6 +145,16 @@ Curated language templates:
+
+The templates store has templates from openfaas and openfaas-incubator organizations with openfaas templates being the official ones.
+
+In order to see what templates are available in the store use `faas-cli template store list` and filter openfaas or openfaas-incubator organizations with `--source` flag.
```
Remove the source flag and mention here

martindekov replied:
source removed ✔️

martindekov commented:
updated examples and PR description, the `templates.json` still uses `cli-lang` instead of `template` as key so the PR still uses that key to unmarshall, when PR is merged I will amend that

alexellis commented:
For the `--url` parameter should we expose an environmental variable like `OPENFAAS_TEMPLATES_URL`? Therefore if you work somewhere like @LucasRoesler's company you can set this for developers and they no longer have to keep typing in `--url some-long-url` for every command?

alexellis commented:
If you think this is a good idea, raise an issue for after the merge of this PR.

martindekov commented:
Yeah it is totally a good idea, I saw it in pull command, can be replicated here also

martindekov commented:
Amended the change `cli-lang` -> `template`

martindekov commented:
Ok amended the change and added the following instruction to the PR also added test to cover the function and showcase the priority:

Also you can set the environmental variable OPENFAAS_TEMPLATE_STORE_URL which will prioritize over the default one the following way `argument` > `environmental_variable` > `default`

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,153 @@
+	}
+
+	client := http.DefaultClient
+	res, clientErr := client.Do(req)
```
Do we want to add a timeout to this request?

One way to do this is 

```golang
ctx, cancel := context.WithTimeout(req.Context(), 5*time.Second)
defer cancel()

req = req.WithContext(ctx)
```

alexellis replied:
Yes a timeout makes sense.

martindekov replied:
Ok adding it thanks for showing me how to do it

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,153 @@
+		return nil, fmt.Errorf("error while requesting template list: %s", clientErr.Error())
+	}
+
+	if res != nil {
```
If `clientErr==nil` can also be `res ==nil`?  Not sure if we need this check after checking the error

martindekov replied:
Yep I will add the changes in different commit and squash if needed so the history won't be overwritten

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,153 @@
+	return nil, fmt.Errorf("error empty response from: %s", templateStoreURL)
+}
+
+func formatCommandOutput(templates []TemplateInfo) string {
```
What is the benefit of wrapping the call to `formatTemplatesOutput` here?  Why not call it directly?

martindekov replied:
I used to have logic which was removed, I will remove the wrapper

martindekov replied:
Done ✔️

LucasRoesler commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,69 @@
+	var nonExistingTemplates []string
+	if len(args) == 0 {
+		return fmt.Errorf("\nNeed to specify one of the store templates check available ones by running the command:\n\nfaas-cli template store list\n")
+	} else {
```
You can reduce the indentation level here, we don't need this else block because the if `return`s an error

martindekov replied:
Yeah ... totally missed that out

martindekov replied:
Yep ✔️

LucasRoesler commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,153 @@
+			}
+			return templatesInfo, nil
+		}
+		return nil, fmt.Errorf("error empty response body from: %s", templateStoreURL)
```
We could also reduce the indentation level here by writing this as 

```golang
if res.Body == nil {
     return nil, fmt.Errorf("error empty response body from: %s", templateStoreURL)
}
defer res.Body.Close()
.
.
.
```

martindekov replied:
Good point ^

martindekov replied:
Done ✔️

LucasRoesler commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,69 @@
+			for _, storeTemplate := range storeTemplates {
+				sourceName := fmt.Sprintf("%s/%s", storeTemplate.Source, storeTemplate.TemplateName)
+				if template == storeTemplate.TemplateName || template == sourceName {
+					runTemplatePull(cmd, []string{storeTemplate.Repository})
```
`runTemplatePull` can return errors, those should be caught and handled here

martindekov replied:
Ah forgot about changing that, will amend

martindekov replied:
Error handler ✔️

LucasRoesler commented on commands/template_pull.go:
```diff
@@ -20,74 +20,42 @@ var (
-	},
-	Short: "Downloads templates from the specified github repo",
+	Use:   `pull [REPOSITORY_URL]`,
+	Short: `Downloads templates from the specified github repo`,
```
I don't see where the check for at least one arg went after you refactored this.  Shouldn't it now be 

```golang 
Args: cobra.MinimumNArgs(1),
```

martindekov replied:
I have logic for checking the first argument (as url) inside the function itself, the (1) argument had to be there since the command was `template pull` and `pull` was necessary and now since we have separated those commands into two (`pull`) it needs only 1 argument arg[0] to be valid. I fixed the check in here:

```
...
func runTemplatePull(cmd *cobra.Command, args []string) error {
	repository := ""
	if len(args) > 0 {
		repository = args[0]
	}
...
```

martindekov replied:
I handled this problem I hope that is what you meant @LucasRoesler . I really appreciate your feedback thanks!

alexellis commented on README.md:
```diff
@@ -145,6 +145,18 @@ Curated language templates:
+
+> Note: You can set your own custom store location with `--url` flag or set `OPENFAAS_TEMPLATE_STORE_URL` environmental variable
+
+To pull templates from the store just write the name of the template you want, you can pull couple of templates at once `faas-cli template store pull go golang-http python3-debian ...`
```
I didn't mention multiple-pulls in the same command in the original issue. I think we could probably reduce complexity by removing this behavior.

The main use-case is for templates to be pulled from the default location or a special signal template repo to be pulled in for a project.

martindekov replied:
Ok yeah I will remove the feature

martindekov replied:
Feature removed ✔️

alexellis commented on commands/priority.go:
```diff
@@ -9,8 +9,9 @@ import (
-	templateURLEnvironment = "OPENFAAS_TEMPLATE_URL"
+	openFaaSURLEnvironment      = "OPENFAAS_URL"
+	templateURLEnvironment      = "OPENFAAS_TEMPLATE_URL"
+	templateStoreURLEnvironment = "OPENFAAS_TEMPLATE_STORE_URL"
```
Please can you document each of these three env-vars in the docs and in this repo's README file? Happy for you to open a new issue to cover that work.

martindekov replied:
Will probably open a new Issue for this

martindekov replied:
Issue as pointed ^

martindekov replied:
Actually all three of those are documented in `README.md`

martindekov replied:
I will open an issue for `openfaas/docs`

martindekov replied:
ok added documentation in `TEMPLATES.md` on the new feature also

alexellis commented on commands/template.go:
```diff
@@ -0,0 +1,24 @@
+	Example: `  faas-cli template pull https://github.com/custom/template
+  faas-cli template store list
+  faas-cli template store ls
+  faas-cli template store pull ruby-http
```
The docs say that multiple template/repo names are possible here, but the help text doesn't represent that same data.

I'd prefer us to only allow one repo/template to be pulled at a time using the CLI.

martindekov replied:
ok I will remove that logic but should I do this in new commit, I think I can force push in this one ?

LucasRoesler replied:
A new commit addressing a single comment is preferable because it is easier to then track. Then rebase and squash once at the end after it has been fully approved.

martindekov replied:
Ok thanks again @LucasRoesler I will proceed with that approach from now on

martindekov replied:
Removed the feature ✔️

alexellis commented on commands/template_store_list.go:
```diff
@@ -0,0 +1,151 @@
+	Aliases: []string{"ls"},
+	Long:    `List templates from official store or from custom URL`,
+	Example: `  faas-cli template store list
+  faas-cli template store ls
```
@johnmccabe should we mention the ENV-VAR in the command help text which can be used to override the template store URL?

martindekov replied:
Mentioned it in `Long` explanation not in `Example` ✔️

alexellis commented on commands/template_store_pull.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// templateStorePullCmd pulls templates from default store or custom store if set
+var templateStorePullCmd = &cobra.Command{
```
For future reference - this is generally only needed on public / exported types.

alexellis commented on guide/TEMPLATE.md:
````diff
@@ -65,3 +65,31 @@ faas-cli template pull
 faas-cli new --list
 ```
+
+## Check template store
````
Can we get a PR open for openfaas/docs please?
